### PR TITLE
New NNFrames interface

### PR DESCRIPTION
**Pull request by hhbyyh**
Following interface are supported for both scala and python:
```
      NNEstimator(model, criterion),
      NNEstimator(model, criterion, Array(6), Array(1)),
      NNEstimator(model, criterion, SeqToTensor(Array(6)), SeqToTensor(Array(1)))

      NNModel(model),
      NNModel(model, Array(6)),
      NNModel(model, SeqToTensor(Array(6)))

      NNClassifier(model, criterion),
      NNClassifier(model, criterion, Array(6)),
      NNClassifier(model, criterion, SeqToTensor(Array(6)))

      NNClassifierModel(model),
      NNClassifierModel(model, Array(6)),
      NNClassifierModel(model, SeqToTensor(Array(6)))
```
Scala NNEstimator also support `NNEstimator(model, criterion, FeatureTosample)`, python does not have the equivalent. Python users would need to invoke setSamplePreprocessing.

Meanwhile, for advanced use case (model with multiple input tensor), NNEstimator and NNClassifier support:
     `setSamplePreprocessing(value: Preprocessing[(Any, Option[Any]), Sample[T]]):`

NNModel and NNClassifierModel support:
     `setSamplePreprocessing(value: Preprocessing[Any, Sample[T]])`

For Scala, constructor of `NNEstimator, NNModel, NNClassifier, NNClassifierModel` are set to private, user should alway create instance via` apply` methods.

### CODE REVIEW

**Comment by hhbyyh** [comment]
cc @dding3 @jason-dai for review. Thanks

File: pyzoo/zoo/pipeline/nnframes/nn_classifier.py [hunk 1]
```diff
@@ -296,60 +291,43 @@ class NNModel(JavaTransformer, HasFeaturesCol, HasPredictionCol, HasBatchSize, J
...
-               to a Tensor.
-               ScalarToTensor transform a number to a Tensor with single dimension of length 1.
-               Multiple Preprocessing can be combined as a ChainedPreprocessing.
-        :param jvalue: Java object create by Py4j
-        :param bigdl_type(optional): Data type of BigDL model, "float"(default) or "double".
-        :return:
-        """
-        chained_transformer = ChainedPreprocessing([feature_preprocessing, TensorToSample()])
-        return NNModel(model, chained_transformer, jvalue, bigdl_type)
+        self._setDefault(batchSize=1)
```
[the remark this answers is not in the document]
Comment (hhbyyh) [reply pyzoo/zoo/pipeline/nnframes/nn_classifier.py]: Thanks. will remove it.

**Comment by hhbyyh** [comment]
Updated. Thanks.

**Comment by hhbyyh** [comment]
http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/297/

**Comment by dding3** [comment]
LGTM

**Comment by jason-dai** [comment]
LGTM

**Comment by hhbyyh** [comment]
Thanks for review. @jenniew Please let me if there's concern about the private constructor. I can send follow-up. Merging to master.

**Pull request merged**